<a href="https://colab.research.google.com/github/ai6ph/Loan-default-prediction/blob/main/loan_default_prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Objective of this project is to predict if the loanee will be a defaulter or not. 
We will use the custom dataset available on the Kaggle. 
We will use the simple and basic Data Science and machine learning methods to execute the solution.

In [ ]:
#Including the basic libraries at the start
import pandas as pd
import numpy as np

Now let us impot the dataset to our notebook

In [ ]:
#importing the dataset
df=pd.read_csv("/content/Default_Fin.csv")

Now let us have a simple look at our data.

In [ ]:
df.head()

Index  Employed  Bank Balance  Annual Salary  Defaulted?
0      1         1       8754.36      532339.56           0
1      2         0       9806.16      145273.56           0
2      3         1      12882.60      381205.68           0
3      4         1       6351.00      428453.88           0
4      5         1       9427.92      461562.00           0

Now before we move ahead and start using the dataset, let us see if there is any need of data pre-processing by counting null or NaN values if any.

In [ ]:
df.isnull().sum()

Index            0
Employed         0
Bank Balance     0
Annual Salary    0
Defaulted?       0
dtype: int64

So, we can clearly see that there is no NaN values present in our data, we can directly move ahead and see the interelation of our features present in dataset.

As we know that our target variable is "Defaulted?", we will look at the correlation present amoung all of them so we can see it there are any features we don't need in model development.

In [ ]:
#looking at the inter-relation or correlation amoung all of the features
df.corr()

Index  Employed  Bank Balance  Annual Salary  Defaulted?
Index          1.000000 -0.002597      0.010342      -0.001427   -0.005135
Employed      -0.002597  1.000000     -0.203578       0.753985   -0.035420
Bank Balance   0.010342 -0.203578      1.000000      -0.152243    0.350119
Annual Salary -0.001427  0.753985     -0.152243       1.000000   -0.019871
Defaulted?    -0.005135 -0.035420      0.350119      -0.019871    1.000000

In [ ]:
#Dropping "Defaulted?" and taking remaining as independent variable X
X = df.drop(['Defaulted?'], axis="columns")

#Taking "Defaulted?" as dependent variable Y
Y = df["Defaulted?"]

Now, as we have only one dataset, we will split it into 2 parts, named test and train dataset.
This will help us in giving more clear approach at model training and testing.

In [ ]:
#Importing required libraries for splitting dataset 
from sklearn.model_selection import train_test_split

#Diving dataset to X_train and Y_train as training variables and y_train and y_test as testing variables
#Making the ratio of 70% training and 30% testing as it is most efficient ratio.

x_train, x_test, y_train, y_test = train_test_split(X,Y, test_size = 0.30, random_state = 0)

Once we have splitted the dataset, we will import the Logistic Regression into our notebook, as it only deals with 2 values, 0 or 1 and True or False which is the best fit for this case.

In [ ]:
#Importing the Logistic Regression 
from sklearn.linear_model import LogisticRegression

#Creating Logistic Regression object
lr = LogisticRegression()


Now we will just train our model using x_train and y_train variables which we got by splitting dataset.

In [ ]:
#Fitting model with x_train and y_train variables
lr.fit(x_train,y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

Now we will predict the values over testing dataset and see what happens.

In [ ]:
Yhat=lr.predict(x_test)
Yhat

array([0, 0, 0, ..., 0, 0, 0])

Now finally, let us see our model accuracy firstly on training datset

In [ ]:
lr.score(x_train,y_train)

0.9662857142857143

and finally accuracy of our model for testing dataset

In [ ]:
lr.score(x_test,y_test)

0.961

We can conclude that, we have successfully built a Machine Learning model to predict the Loan Defaulters using a custom dataset with a training accuracy of 96.6% approximately and testing accuracy of 96.1%.